src/lib/formats/rpk.cpp

```diff
@@ -0,0 +1,459 @@
+	return std::error_condition();
+}
+
+
+//-------------------------------------------------
+//  add_ram_socket
+//-------------------------------------------------
+
+std::error_condition util::rpk_file::add_ram_socket(std::string &&id, const util::xml::data_node &ram_resource_node)
+{
+	// find the length attribute
+	std::string const *const length_string = ram_resource_node.get_attribute_string_ptr("length");
+	if (!length_string)
+		return rpk_reader::error::MISSING_RAM_LENGTH;
+
+	// parse it
+	unsigned int length;
+	char suffix = '\0';
+	sscanf(length_string->c_str(), "%u%c", &length, &suffix);
+	switch (tolower(suffix))
+	{
+	case 'k':
+		// kilobytes
+		length *= 1024;
+		break;
+
+	case 'm':
+		// megabytes
+		length *= 1024 * 1024;
+		break;
+
+	case '\0':
+		break;
+
+	default:  // failed
+		return rpk_reader::error::INVALID_RAM_SPEC;
+	}
+
+	// determine the type of RAM
+	rpk_socket::socket_type type;
+	std::string const *const ram_type = ram_resource_node.get_attribute_string_ptr("type");
+	if (ram_type && *ram_type == "persistent")
+		type = rpk_socket::socket_type::PERSISTENT_RAM;
+	else
+		type = rpk_socket::socket_type::RAM;
+
+	// persistent RAM needs a file name
+	std::string file;
+	if (type == rpk_socket::socket_type::PERSISTENT_RAM)
+	{
+		std::string const *const ram_filename = ram_resource_node.get_attribute_string_ptr("file");
+		if (ram_filename == nullptr)
+			return rpk_reader::error::INVALID_RAM_SPEC; // <ram type='persistent'> must have a 'file' attribute
+		file = *ram_filename;
+	}
+
+	// finally add the socket
+	m_sockets.emplace_back(*this, std::move(id), type, std::move(file), std::nullopt, length);
+	return std::error_condition();
+}
+
+
+/***************************************************************************
+	RPK SOCKET
+***************************************************************************/
+
+//-------------------------------------------------
+//  ctor
+//-------------------------------------------------
+
+util::rpk_socket::rpk_socket(rpk_file &rpk, std::string &&id, socket_type type, std::string &&filename, std::optional<hash_collection> &&hashes, std::optional<std::uint32_t> length)
+	: m_rpk(rpk)
+	, m_id(id)
+	, m_type(type)
+	, m_filename(filename)
+	, m_hashes(std::move(hashes))
+	, m_length(length)
+{
+}
+
+
+//-------------------------------------------------
+//  dtor
+//-------------------------------------------------
+
+util::rpk_socket::~rpk_socket()
+{
+}
+
+
+//-------------------------------------------------
+//  read_file
+//-------------------------------------------------
+
+std::error_condition util::rpk_socket::read_file(std::vector<std::uint8_t> &result) const
+{
+	// find the file
+	if (m_rpk.zipfile().search(m_filename, false) < 0)
+		return rpk_reader::error::INVALID_FILE_REF;
+
+	// prepare a buffer
+	result.resize(m_rpk.zipfile().current_uncompressed_length());
```

Comment:
If you’re using a vector, you should clear it before resizing to avoid a copy if it already contained something (clearing doesn’t deallocate), and you should catch the potential `std::bad_alloc` to report `std::errc::not_enough_memory`.